common/src/test/java/io/netty/util/RecyclerTest.java

```diff
@@ -34,6 +36,43 @@ protected HandledObject newObject(
         };
     }
 
+    @Test(timeout = 5000L)
+    public void testThreadCanBeCollectedEvenIfHandledObjectIsReferenced() throws Exception {
+        final Recycler<HandledObject> recycler = newRecycler(1024);
+        final AtomicBoolean collected = new AtomicBoolean();
+        final AtomicReference<HandledObject> reference = new AtomicReference<HandledObject>();
+        Thread thread = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                HandledObject object = recycler.get();
+                // Store a reference to the HandledObject to ensure it is not collected when the run method finish.
+                reference.set(object);
+            }
+        }) {
+            @Override
+            protected void finalize() throws Throwable {
+                super.finalize();
+                collected.set(true);
+            }
+        };
+        assertFalse(collected.get());
+        thread.start();
+        thread.join();
+
+        // Null out so it can be collected.
+        thread = null;
+
+        // Loop until the Thread was collected. If we can not collect it the Test will fail due of a timeout.
+        while (!collected.get()) {
```

Comment:
This test is a bit speculative, and dependent upon VM/GC behavior. Hopefully all GCs behaves as expected and will reclaim/finalize the object in time. I'm not sure of a better way to test this, but this maybe subject to sporadic test failures (maybe unlikely though?).

Reply:
@Scottmitch I think its the best we can do... If there will be ever a better API I am happy to revisit it.

Reply:
If it starts failing I guess we can take a closer look and consider disabling it if necessary.